iavl/iavl.py

```diff
@@ -394,32 +394,36 @@ def remove_recursive(
 ) -> (Optional[bytes], Optional[NodeRef]):
     """
     returns (removed value, new node if changed)
-    - (None, _) -> nothing changed in subtree
-    - (value, None) -> leaf node is removed
-    - (value, new node) -> subtree changed
+    - (None, _, None) -> nothing changed in subtree
+    - (value, None, newKey) -> leaf node is removed
+    - (value, new node, newKey) -> subtree changed
     """
     node = ndb.resolve_node(ref)
     if node.is_leaf():
         if node.key == key:
-            return node.value, None
+            return node.value, None, None
         else:
-            return None, None
+            return None, None, None
     else:
         turn_left = key < node.key
         if turn_left:
-            value, new_child = remove_recursive(ndb, key, node.left_node_ref, version)
+            value, new_child, new_key = remove_recursive(
+                ndb, key, node.left_node_ref, version
+            )
         else:
-            value, new_child = remove_recursive(ndb, key, node.right_node_ref, version)
+            value, new_child, new_key = remove_recursive(
+                ndb, key, node.right_node_ref, version
+            )
 
         if value is None:
-            return None, None
+            return None, None, None
 
         if new_child is None:
             # return the other child
             if turn_left:
-                return value, node.right_node_ref
+                return value, node.right_node_ref, node.key
```

Comment:
not sure why only return newKey here while other return None

Reply:
haha, very tricky

Reply:
The key here is to maintain invariant: `node.key == minimal_key in node.right subtree`